### PR TITLE
return 401 if source origin is invalid

### PR DESCRIPTION
**Pull request by sebastianbenz**
Wasn't aware that the origin is optional. Verifying the request host in this case + check for localhost in dev mode.

### CODE REVIEW

**Comment by dvoytenko** [comment]
@sebastianbenz would do you mean "origin is optional"?

File: middlewares/amp-access-cors.js [hunk 1]
```diff
@@ -62,16 +60,27 @@ module.exports = function(req, res, next) {
...
  * - *.ampproject.org
  * - The Publisher’s own origins
+ *
  */
-function isValidOrigin(origin) {
-  return matchesAnyOrigin(VALID_ORIGINS, origin) || isValidSourceOrigin(origin);
+function isValidOrigin(req, origin) {
+  if (!origin) {
+    // if origin not set verify the request host
+    origin = req.protocol + '://' + req.hostname;
```
Comment (dvoytenko) [inline middlewares/amp-access-cors.js]: `req.protocol` does not contain `:` already? Seems a bit unusual for nodejs.

File: middlewares/amp-access-cors.js [hunk 2]
```diff
@@ -84,6 +93,9 @@ function matchesAnyOrigin(validOrigins, origin) {
     return false;
   }
   var host = url.parse(origin).hostname;
+  if(!host) {
```
Comment (dvoytenko) [inline middlewares/amp-access-cors.js]: nit: space

Comment (sebastianbenz) [inline middlewares/amp-access-cors.js]: done

File: middlewares/amp-access-cors.js (hunk 1 above)
Comment (sebastianbenz) [inline middlewares/amp-access-cors.js]: seems like nodejs is sometimes unusual.

**Comment by sebastianbenz** [comment]
PTAL

**Comment by dvoytenko** [comment]
LGTM

**Comment by sebastianbenz** [comment]
@dvoytenko by optional I meant that the origin isn't submitted with every request, but only for CORS requests.

**Pull request merged**